source/common/network/address_impl.cc

```diff
@@ -64,6 +64,62 @@ int Ipv4Instance::socket(SocketType type) const {
   return ::socket(AF_INET, flagsFromSocketType(type), 0);
 }
 
+std::array<uint8_t, 16> Ipv6Instance::Ipv6Helper::address() const {
+  std::array<uint8_t, 16> result;
+  std::copy(std::begin(address_.sin6_addr.s6_addr), std::end(address_.sin6_addr.s6_addr),
+            std::begin(result));
+  return result;
+}
+
+uint32_t Ipv6Instance::Ipv6Helper::port() const { return ntohs(address_.sin6_port); }
+
+std::string Ipv6Instance::Ipv6Helper::makeFriendlyAddress() const {
+  char str[INET6_ADDRSTRLEN];
+  const char* ptr = inet_ntop(AF_INET6, &address_.sin6_addr, str, INET6_ADDRSTRLEN);
+  ASSERT(str == ptr);
+  return ptr;
+}
+
+Ipv6Instance::Ipv6Instance(const sockaddr_in6& address) : InstanceBase(Type::Ip) {
+  ip_.ipv6_.address_ = address;
+  ip_.friendly_address_ = ip_.ipv6_.makeFriendlyAddress();
+  friendly_name_ = fmt::format("[{}]:{}", ip_.friendly_address_, ip_.port());
+}
+
+Ipv6Instance::Ipv6Instance(const std::string& address) : Ipv6Instance(address, 0) {}
+
+Ipv6Instance::Ipv6Instance(const std::string& address, uint32_t port) : InstanceBase(Type::Ip) {
+  memset(&ip_.ipv6_.address_, 0, sizeof(ip_.ipv6_.address_));
+  ip_.ipv6_.address_.sin6_family = AF_INET;
+  ip_.ipv6_.address_.sin6_port = htons(port);
+  if (!address.empty()) {
+    if (1 != inet_pton(AF_INET6, address.c_str(), &ip_.ipv6_.address_.sin6_addr)) {
+      throw EnvoyException(fmt::format("invalid ipv6 address '{}'", address));
+    }
+  } else {
+    ip_.ipv6_.address_.sin6_addr = in6addr_any;
+  }
+  // Just in case address is in a non-canonical format, format from network address.
+  ip_.friendly_address_ = ip_.ipv6_.makeFriendlyAddress();
+  friendly_name_ = fmt::format("[{}]:{}", ip_.friendly_address_, ip_.port());
+}
+
+Ipv6Instance::Ipv6Instance(uint32_t port) : Ipv6Instance("", port) {}
+
+int Ipv6Instance::bind(int fd) const {
+  return ::bind(fd, reinterpret_cast<const sockaddr*>(&ip_.ipv6_.address_),
+                sizeof(ip_.ipv6_.address_));
+}
+
+int Ipv6Instance::connect(int fd) const {
+  return ::connect(fd, reinterpret_cast<const sockaddr*>(&ip_.ipv6_.address_),
+                   sizeof(ip_.ipv6_.address_));
+}
+
+int Ipv6Instance::socket(SocketType type) const {
+  return ::socket(AF_INET, flagsFromSocketType(type), 0);
```

Comment:
@jamessynge sorry I didn't notice this before, but this should be AF_INET6. (I'm sure you would have noticed that when you start using this). In follow ups can you make sure that there is full coverage for the new code? There is a bunch missing now.